### PR TITLE
bpo-29591: Upgrade Modules/expat to libexpat 2.2

### PR DESCRIPTION
**Pull request by vstinner**
[Security] bpo-29591: Update expat copy from 2.1.1 to 2.2.0 to get fixes of CVE-2016-0718 and CVE-2016-4472. See https://sourceforge.net/p/expat/bugs/537/ for more information.

### CODE REVIEW

File: Modules/expat/expat.h [hunk 1]
```diff
@@ -342,7 +342,7 @@ XML_SetEntityDeclHandler(XML_Parser parser,
                          XML_EntityDeclHandler handler);
 
 /* OBSOLETE -- OBSOLETE -- OBSOLETE
-   This handler has been superseded by the EntityDeclHandler above.
+   This handler has been superceded by the EntityDeclHandler above.
```
Comment (vstinner) [inline Modules/expat/expat.h]: wait, superseded is the right spelling no?

File: Modules/expat/amigaconfig.h [hunk 2]
```diff
@@ -7,6 +7,9 @@
 /* Define to 1 if you have the `bcopy' function. */
 #define HAVE_BCOPY 1
 
+/* Define to 1 if you have the <check.h> header file. */
+#undef HAVE_CHECK_H
```
Comment (vstinner) [inline Modules/expat/amigaconfig.h]: I don't understand this change, I just tried to reduce the changes.

File: Modules/expat/expat_external.h [hunk 3]
```diff
@@ -7,6 +7,10 @@
 
 /* External API definitions */
 
+/* Namespace external symbols to allow multiple libexpat version to
+   co-exist. */
+#include "pyexpatns.h"
```
Comment (vstinner) [inline Modules/expat/expat_external.h]: I would be nice to upstream pyexpatns.h namespace, it seems useful to build a clean DLL on Windows.

File: Modules/expat/expat.h (hunk 1 above)
Comment (gpshead) [reply Modules/expat/expat.h]: it is... but lets not correct spelling in third party code.

Comment (gpshead) [reply Modules/expat/expat.h]: oh, i see, this was a python tree only change?  weird.  keep it as superseded.

File: Modules/expat/amigaconfig.h (hunk 2 above)
Comment (gpshead) [reply Modules/expat/amigaconfig.h]: we don't even need to carry this file around given the `#elif defined(__amigaos__)` it is included under.

Comment (vstinner) [reply Modules/expat/amigaconfig.h]: I'm not sure that I understood. Do you want to remove amigaconfig.h?

Comment (vstinner) [reply Modules/expat/amigaconfig.h]: In the meanwhile, I reverted this change.

File: Modules/expat/expat.h (hunk 1 above)
Comment (vstinner) [reply Modules/expat/expat.h]: > it is... but lets not correct spelling in third party code.

Right, I prefer to reduce the downstream changes: so I reverted this change as well.

**Review by gpshead — DISMISSED** [review]
I believe we can drop amigaconfig.h, watcomconfig.h, and macconfig.h (used for MacOS classic on PPC32).

**Comment by matrixise** [comment]
Duplicate of https://github.com/python/cpython/pull/2021 ?

**Comment by vstinner** [comment]
@matrixise: "Duplicate of #2021 ?"

I chose a different approach. I wanted to recreate Modules/expat/ directory "from scratch", so I wrote a shell script that I attached http://bugs.python.org/issue29591

I tried to restore Python changes reverted by my approach, to more easily identify downstream (Python) changes.

I compare my PR with your (rebased) PR:

* you didn't update COPYING
* you removed: #include "pyexpatns.h"
* you removed: #define XML_HAS_SET_HASH_SALT
* my PR now also removes amigaconfig.h, macconfig.h and watcomconfig.h of Modules/expat/

**Comment by gpshead** [comment]
FWIW I agree with recreating the directory from scratch and applying the few actually relevant local changes on top as you have done.

**Comment by gpshead** [comment]
We can remove the XML_HAS_SET_HASH_SALT define and remove that from the ifdef in Modules/pyexpat.c as that was a custom Python tree only addition I did back when I backported those changes to the old version of expat we had at the time.

**Comment by vstinner** [comment]
@gpshead: "We can remove the XML_HAS_SET_HASH_SALT define and remove that from the ifdef in Modules/pyexpat.c (...)"

Right, done.

**Comment by vstinner** [comment]
Does it look good to you now @gpshead?

**Review by tiran — APPROVED** [review]
Your patch on top of expat upstream looks good to me. I don't have enough resources to properly verify all changes in expat, though.

**Comment by vstinner** [comment]
Thanks for the reviews @tiran and @gpshead!

**Pull request merged**